libbeat/processors/actions/alterFieldProcessor.go

```diff
@@ -0,0 +1,175 @@
+	config := struct {
+		Fields        []string `config:"fields"`
+		IgnoreMissing bool     `config:"ignore_missing"`
+		FailOnError   bool     `config:"fail_on_error"`
+		FullPath      bool     `config:"full_path"`
+	}{
+		IgnoreMissing: false,
+		FailOnError:   true,
+		FullPath:      false,
+	}
+
+	if err := c.Unpack(&config); err != nil {
+		return nil, fmt.Errorf("failed to unpack the %s fields configuration: %s", processorName, err)
+	}
+
+	// Skip mandatory fields
+	for _, readOnly := range processors.MandatoryExportedFields {
+		for i, field := range config.Fields {
+			if field == readOnly {
+				config.Fields = append(config.Fields[:i], config.Fields[i+1:]...)
+			}
+		}
+	}
+
+	return &alterFieldProcessor{
+		Fields:        config.Fields,
+		IgnoreMissing: config.IgnoreMissing,
+		FailOnError:   config.FailOnError,
+		processorName: processorName,
+		alterFunc:     alterFunc,
+	}, nil
+
+}
+
+func (a *alterFieldProcessor) String() string {
+	return fmt.Sprintf("%s fields=%+v", a.processorName, *a)
+}
+
+func (a *alterFieldProcessor) Run(event *beat.Event) (*beat.Event, error) {
+	var backup *beat.Event
+	if a.FailOnError {
+		backup = event.Clone()
+	}
+
+	for _, field := range a.Fields {
+		err := a.alter(event, field)
+		if err != nil {
+			if a.FailOnError {
+				event = backup
+				event.PutValue("error.message", err.Error())
+				return event, err
+			}
+		}
+	}
+
+	return event, nil
+}
+
+func (a *alterFieldProcessor) alter(event *beat.Event, field string) error {
+	// Get the value of the field to alter
+
+	key, value := getcaseInsensitiveValue(event.Fields, field)
+	if value == nil {
+		if a.IgnoreMissing {
+			return nil
+		}
+		return fmt.Errorf("could not fetch value for key: %s, Error: %v", field, mapstr.ErrKeyNotFound)
+	}
+
+	// Delete the exisiting value
+	if err := event.Delete(key); err != nil {
+		return fmt.Errorf("could not delete key: %s, Error: %v", key, err)
+	}
+
+	// Alter the field
+	var alterString string
+	if strings.ContainsRune(key, '.') {
+		// In case of nested fields provided, we need to make sure to only modify the latest key in the chain
+		lastIndexRuneFunc := func(r rune) bool { return r == '.' }
+		idx := strings.LastIndexFunc(key, lastIndexRuneFunc)
+		alterString = key[:idx+1] + a.alterFunc(key[idx+1:])
+	} else {
+		alterString = a.alterFunc(key)
+	}
+
+	// Put the field back
+	if _, err := event.PutValue(alterString, value); err != nil {
+		return fmt.Errorf("could not put value: %s: %v, Error: %v", alterString, value, err)
+	}
+
+	return nil
+}
+
+func getcaseInsensitiveValue(event mapstr.M, field string) (key string, value interface{}) {
+
+	// Fast path, key is present as is.
+	if v, err := event.GetValue(field); err == nil {
+		return field, v
+	}
+
+	// iterate through the map for case insensitive search
+	fieldList := strings.Split(field, ".")
+	data := event
+
+	// outer function goes through the field keyword seperated by '.'
+	for i, subKey := range fieldList {
```

Comment:
Unfortunately, I don't think this approach would work correctly. Or I might have missed something.

The events can have this form too:

```
{
  first: {
    second: "value
    second.third: {
      fourth: "value"
    }
  }
}
```

1. A key on each level can contain `.`
2. A value does not have to be an object, it can be a primitive type.

I suggest you include these into your test cases.

Reply:
Never mind about 1. The original `mapstr.M` implementation does not support keys with `.` in them.